### ISSUE TITLE
AmazonClientException on CognitoIdentityProvider/CognitoUser.confirmPasswordInBackground

### ISSUE THREAD

**Issue by Knight704** [aws/aws-sdk-android#191]
Hello, I experienced crash implementing forgot password feature.
Looks like that on the second step, where I have to enter confirmation code with a new password
ForgotPasswordContinuation lead to CognitoUser.confirmPasswordInBackground that at the moment
catches only AmazonServiceException.

In my case I got AmazonClientException (so it crashes inside SDK). The way I got this: switch-off internet before second step and continue.
callstack:

```
Fatal Exception: com.amazonaws.AmazonClientException
Unable to execute HTTP request: Unable to resolve host "cognito-idp.us-east-1.amazonaws.com": No address associated with hostname

com.amazonaws.http.AmazonHttpClient.executeHelper (AmazonHttpClient.java:424)
com.amazonaws.http.AmazonHttpClient.execute (AmazonHttpClient.java:199)
com.amazonaws.services.cognitoidentityprovider.AmazonCognitoIdentityProviderClient.invoke (AmazonCognitoIdentityProviderClient.java:1918)
com.amazonaws.services.cognitoidentityprovider.AmazonCognitoIdentityProviderClient.confirmForgotPassword (AmazonCognitoIdentityProviderClient.java:637)
com.amazonaws.mobileconnectors.cognitoidentityprovider.CognitoUser.confirmPasswordInternal (CognitoUser.java:524)
com.amazonaws.mobileconnectors.cognitoidentityprovider.CognitoUser.access$000 (CognitoUser.java:103)
com.amazonaws.mobileconnectors.cognitoidentityprovider.CognitoUser$4.run (CognitoUser.java:465)
java.lang.Thread.run (Thread.java:841)
Caused by java.net.UnknownHostException
java.net.InetAddress.lookupHostByName (InetAddress.java:394)
libcore.net.http.HttpsURLConnectionImpl.getOutputStream (HttpsURLConnectionImpl.java:281)
com.amazonaws.http.UrlHttpClient.writeContentToConnection (UrlHttpClient.java:128)
com.amazonaws.http.UrlHttpClient.execute (UrlHttpClient.java:65)
com.amazonaws.http.AmazonHttpClient.executeHelper (AmazonHttpClient.java:356)
com.amazonaws.http.AmazonHttpClient.execute (AmazonHttpClient.java:199)
com.amazonaws.services.cognitoidentityprovider.AmazonCognitoIdentityProviderClient.invoke (AmazonCognitoIdentityProviderClient.java:1918)
com.amazonaws.services.cognitoidentityprovider.AmazonCognitoIdentityProviderClient.confirmForgotPassword (AmazonCognitoIdentityProviderClient.java:637)
com.amazonaws.mobileconnectors.cognitoidentityprovider.CognitoUser.confirmPasswordInternal (CognitoUser.java:524)
com.amazonaws.mobileconnectors.cognitoidentityprovider.CognitoUser.access$000 (CognitoUser.java:103)
com.amazonaws.mobileconnectors.cognitoidentityprovider.CognitoUser$4.run (CognitoUser.java:465)
java.lang.Thread.run (Thread.java:841)
```

**Comment by wdane** [comment]
Can you help me understand the concern. Is the concern that there is no way to catch this exception if the user is offline, and it will always crash the app?

**Comment by Knight704** [comment]
yes, exactly.
From client code side there is no way to catch this exception.

**Comment by karthiksaligrama** [comment]
You should be able to catch the AmazonClientException in your try catch block. The AmazonClientException is a publicly visible exception.

**Comment by Knight704** [comment]
@karthiksaligrama, no, and this is the point. SDK internally do this:
```
try {
    // this could throw AmazonClientException, which isn't catched by catch block below.
} catch (AmazonServiceException e) {
    // forward exception to listener
}
```

Another thing is worth mentioning that above code is wrapped inside anonymous thread, so client code has no chance to cope with this.
see [https://github.com/aws/aws-sdk-android/blob/master/aws-android-sdk-cognitoidentityprovider/src/main/java/com/amazonaws/mobileconnectors/cognitoidentityprovider/CognitoUser.java#L503](source)

**Comment by wdane** [comment]
@Knight704 From the source you linked to, isn't catching the exception and forwarding to the callback the desired behavior? Just looking at that link it looks like it shouldn't crash the thread and you can handle the exception how you like?

**Comment by Knight704** [comment]
@wdane, of course, it is desired behavior to forward exception to callback. But in case if it would be a AmazonServiceException, thread will crash.
That happened to me when it was AmazonClientException

**Comment by Knight704** [comment]
ok, let me get this straight again,
My client code isn't getting callback with exception, because SDK internally catch only AmazonServiceException, which is not cover all possible exceptions that may happen. And it's crashing **inside SDK**

And, as I can see, this applies not only to ForgotPassword functionality.

**Comment by mreddyg** [comment]
@Knight704 I agree with you. This is a bug in the SDK's exception handling and will be fixed on priority in a future release. Thank you for reporting this bug.

**Comment by georgegatt** [comment]
Any update on this issue ? Has it been fixed ?

**Comment by kvasukib** [comment]
This was fixed in 2.3.9. Please reopen or open a new issue if it still persists.
